### ISSUE TITLE
How can I get the key in time to set the storage path?

### ISSUE THREAD

**Issue by pfrazee** [mafintosh/hyperdrive#146]
In Beaker, I'm storing hyperdrives in a central directory, with each one saved under its public key. If I'm creating a new archive, I don't know the public key prior to calling the `hyperdrive` constructor, so I can't supply the proper path.

From what I can tell, [hypercore generates the key](https://github.com/mafintosh/hypercore/blob/master/index.js#L180) and then [uses the storage for the first time in that flow](https://github.com/mafintosh/hypercore/blob/master/index.js#L196).

So, correct me if I'm wrong, but this should work, right?

```js
var basepath = ...
var archive = hypercore(function (name) {
  return raf(name, {directory: path.join(basepath, archive.key)})
})
```

**Comment by pfrazee** [comment]
```
12:47 PM <jhand> pfrazee: I think https://github.com/mafintosh/deferred-random-access was part of the solution to that
12:47 PM <pfrazee> jhand: yeah but for that to work, I'd need an event I could wait for, right?
12:47 PM <pfrazee> jhand: and the ready event is emitted after storage is used
```

**Comment by pfrazee** [comment]
```
12:51 PM <jhand> pfrazee: the storage function is called before the key is created I think
12:51 PM <jhand> (trying to remember where)
12:52 PM <jhand> pfrazee: ah ya https://github.com/mafintosh/hypercore/blob/917ead034f2195c3d90ea14474a070e41f1e918e/lib/storage.js#L15
12:52 PM <pfrazee> jhand: if that's the case then I need to be able to create the keys prior to calling the constructor, and pass them in, and have hypercore store them
12:52 PM <pfrazee> ah yeah
12:52 PM <pfrazee> yeah then I need to be able to pregenerate the keys and override hypercore's storeSecret setting
```

**Comment by mafintosh** [comment]
to know the public key it first needs to read the public key. i can defer the other ones tho to after that has been loaded. would that work for you?

**Comment by pfrazee** [comment]
In my situation, I'll always know the public key ahead of time, unless I'm creating the archive for the first time. It's just the creation case I need to be able to control. Being able to provide a full keypair, and still have it store the private key, would solve the creation-case for me.

**Comment by mafintosh** [comment]
Fixed by you in mafintosh/hypercore#84

**Comment by pfrazee** [comment]
Also by https://github.com/mafintosh/hyperdrive/pull/147
